overlord/state/change_test.go

```diff
@@ -1119,3 +1119,347 @@ func (ts *taskRunnerSuite) TestCheckTaskDependencies(c *C) {
+	t2.SetToWait(state.DoneStatus)
+	c.Check(chg.Status(), Equals, state.WaitStatus)
+}
+
+func (cs *changeSuite) TestIsWaitingMultipleDependencies(c *C) {
+	st := state.New(nil)
+	st.Lock()
+	defer st.Unlock()
+
+	chg := st.NewChange("change", "...")
+
+	t1 := st.NewTask("task1", "...")
+	t2 := st.NewTask("task2", "...")
+	t3 := st.NewTask("task3", "...")
+	t4 := st.NewTask("wait-task", "...")
+	t3.WaitFor(t1)
+	t3.WaitFor(t2)
+
+	chg.AddTask(t1)
+	chg.AddTask(t2)
+	chg.AddTask(t3)
+	chg.AddTask(t4)
+
+	// Put t4 into wait-status to trigger the isWaiting logic each time
+	// for the change.
+	t4.SetToWait(state.DoneStatus)
+
+	// task1 (do) + task2 (do) => task3 (do) no reboot
+	c.Check(chg.Status(), Equals, state.DoStatus)
+
+	// task1 (done) + task2 (done) => task3 (do) no reboot
+	t1.SetStatus(state.DoneStatus)
+	t2.SetStatus(state.DoneStatus)
+	c.Check(chg.Status(), Equals, state.DoStatus)
+
+	// task1 (done) + task2 (do) => task3 (do) no reboot
+	t1.SetStatus(state.DoneStatus)
+	t2.SetStatus(state.DoStatus)
+	c.Check(chg.Status(), Equals, state.DoStatus)
+
+	// For the next two cases we are testing that a task with dependencies
+	// which have completed, but in a non-successful way is handled correctly.
+	// The reason we expect a DoStatus from the following:
+	// - Task 1 or 2 is in error
+	// - Task 1 or 2 is wait
+	// - Task 3 depends on 1+2, and is in Do.
+	// - Task 4, is independent, and in Wait.
+	// It may seem logically that this change would be in error or wait mode, but because of
+	// our status ordering, Do takes precedence above Wait and Error, so it's important
+	// that our WaitStatus logic (which actually precedes the ordering) is respecting this
+	// by determining the next two cases to be identical, and not opinionated otherwise.
+	// task1 (error) + task2 (wait) => task3 (do) no reboot
+	t1.SetStatus(state.ErrorStatus)
+	t2.SetToWait(state.DoneStatus)
+	c.Check(chg.Status(), Equals, state.DoStatus)
+
+	// task1 (wait) + task2 (error) => task3 (do) no reboot
+	t1.SetToWait(state.DoneStatus)
+	t2.SetStatus(state.ErrorStatus)
+	c.Check(chg.Status(), Equals, state.DoStatus)
+
+	// task1 (done) + task2 (wait) => task3 (do) means need a reboot
+	t1.SetStatus(state.DoneStatus)
+	t2.SetToWait(state.DoneStatus)
+	c.Check(chg.Status(), Equals, state.WaitStatus)
+
+	// task1 (wait) + task2 (wait) => task3 (do) means need a reboot
+	t1.SetToWait(state.DoneStatus)
+	t2.SetToWait(state.DoneStatus)
+	c.Check(chg.Status(), Equals, state.WaitStatus)
+
+	// task1 (done) + task2 (done) => task3 (wait) means need a reboot
+	t1.SetStatus(state.DoneStatus)
+	t2.SetStatus(state.DoneStatus)
+	t3.SetToWait(state.DoneStatus)
+	c.Check(chg.Status(), Equals, state.WaitStatus)
+
+	// task1 (wait) + task2 (abort) => task3 (do)
+	t1.SetToWait(state.DoneStatus)
+	t2.SetStatus(state.AbortStatus)
+	t3.SetStatus(state.DoStatus)
+	c.Check(chg.Status(), Equals, state.AbortStatus)
+}
+
+func (cs *changeSuite) TestIsWaitingUndoTwoTasks(c *C) {
```

Comment:
the Undo tests don't seem completely realistic as in the undo case we are follow the dependencies backward